### PR TITLE
WIP: Making 3.1 the minimum version of CMake

### PR DESCRIPTION
**Pull request by henryiii**
Initial work for #224, making PR early to get Travis build (hoping to fix AMD download).

### CODE REVIEW

**Comment by ddemidov** [comment]
I need to fix OpenCL testing on Travis. It has been broken for a couple of months; it looks like AMD removed the version of OpenCL SDK that was used for the tests.

**Comment by henryiii** [comment]
I have a possible fix, from https://github.com/JuliaGPU/OpenCL.jl

**Comment by henryiii** [comment]
What do you use sudo=True for? It will slow down the build.

**Comment by ddemidov** [comment]
> What do you use sudo=True for? It will slow down the build.

It looks like sudo is not required since f4e2682859a579605fd7f6f9405e5673f44dfa74. Good catch!

**Comment by ddemidov** [comment]
I think my [install_amd_sdk.sh](https://github.com/ddemidov/vexcl/blob/master/.travis/install_amd_sdk.sh) comes from https://github.com/JuliaGPU/OpenCL.jl already. And it does not look like they changed it recently.

**Comment by henryiii** [comment]
Separating out OpenCL fix.

It wasn't different enough... I think the version is the main issue, rather than the agreement page.

**Comment by ddemidov** [comment]
Could you please resolve merge conflicts here?

**Comment by henryiii** [comment]
I need to rebase it and check it. Doing that now.

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,6 +45,14 @@ find_package(Boost REQUIRED COMPONENTS
 add_library(Common INTERFACE)
 add_library(VexCL::Common ALIAS Common)
 
+target_compile_features(Common INTERFACE
```
Comment (ddemidov) [inline CMakeLists.txt]: Is there list of features that cmake knows about? I can see [here](https://cmake.org/cmake/help/v3.1/manual/cmake-compile-features.7.html) that we may want to add `cxx_variadic_templates`.

**Comment by henryiii** [comment]
On my Mac, this is correctly configuring flags now and is correctly avoiding detecting the JIT because of missing OpenMP. I can't build with the Compute backend, either due to updates in Boost (Homebrew always uses the latest, so 1.65.1 I think), or for some other reason, but that's not new to this PR.

File: CMakeLists.txt [hunk 2]
```diff
@@ -134,28 +137,84 @@ endif()
...
+                $<$<CXX_COMPILER_ID:Clang>:${OpenMP_CXX_FLAGS}>
+                $<$<CXX_COMPILER_ID:Intel>:${OpenMP_CXX_FLAGS}>
+                )
+            target_compile_options(JIT INTERFACE ${OpenMP_CXX_FLAGS})
+        endif()
+
+        target_link_libraries(JIT INTERFACE Common ${CMAKE_DL_LIBS})
+        target_compile_definitions(JIT INTERFACE VEXCL_BACKEND_JIT)
 
-    message(STATUS "Found VexCL::JIT")
```
Comment (ddemidov) [inline CMakeLists.txt]: We still report other found backends. Why remove this?

File: CMakeLists.txt (hunk 1 above)
Comment (henryiii) [reply CMakeLists.txt]: Sure, [here](https://cmake.org/cmake/help/v3.1/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES). I think I linked it before somewhere. You can use the dropdown at the top to pick CMake versions.

Comment (ddemidov) [reply CMakeLists.txt]: Thanks! From the top of my head I can only see `cxx_decltype` there.

File: CMakeLists.txt (hunk 2 above)
Comment (henryiii) [reply CMakeLists.txt]: Oops, that would be a mistake.

Comment (henryiii) [reply CMakeLists.txt]: I think it happened when I rebased. Fixed.

**Comment by henryiii** [comment]
Didn't change any code in the last commit, just one line in docs, so skipped ci. Watch the CI build from the commit before.

**Comment by henryiii** [comment]
I believe there was an issue in some versions of CMake where `OpenMP_FOUND` was mentioned in the docs, but `OPENMP_FOUND` was set in the code. I think that's why there was an issue in the last build, rerunning to see if that fixes the problem.

**Comment by henryiii** [comment]
By the way, the `vexcl_add_libraries` helper function is untested (and undocumented) at the moment. `vexcl_add_executables` works in a simple test project.

**Comment by ddemidov** [comment]
Thanks again! I'll look into Boost.Compute issue later, but could you please open an issue for that to make sure we are solving same problem?

**Comment by ddemidov** [comment]
> By the way, the vexcl_add_libraries helper function is untested (and undocumented) at the moment. vexcl_add_executables works in a simple test project.

I like those!

**Comment by ddemidov** [comment]
Boost.Compute backend (with boost-1.65.1) compiles without issues for me on Arch linux. Could your issue be the same as in #222?

**Pull request merged**